lib/widgets/action_sheet.dart

```diff
@@ -1,36 +1,182 @@
-            // TODO: Share raw Markdown, not HTML
-            await Share.shareWithResult(message.content);
-          },
-          child: const Text('Share')),
+        ShareButton(message: message, bottomSheetContext: innerContext, messageListContext: context),
+        if (isComposeBoxOffered) QuoteAndReplyButton(
+          message: message,
+          bottomSheetContext: innerContext,
+          messageListContext: context,
+        ),
       ]);
     });
 }
+
+abstract class MessageActionSheetMenuItemButton extends StatelessWidget {
+  MessageActionSheetMenuItemButton({
+    super.key,
+    required this.message,
+    required this.bottomSheetContext,
+    required this.messageListContext,
+  }) : assert(bottomSheetContext.findAncestorWidgetOfExactType<BottomSheet>() != null),
+       assert(messageListContext.findAncestorWidgetOfExactType<MessageListPage>() != null);
+
+  IconData get icon;
+  String get label;
+  VoidCallback get onPressed;
+
+  final Message message;
+  final BuildContext bottomSheetContext;
+  final BuildContext messageListContext;
+
+  @override
+  Widget build(BuildContext context) {
+    return MenuItemButton(
+      leadingIcon: Icon(icon),
+      onPressed: onPressed,
+      child: Text(label));
+  }
+}
+
+class ShareButton extends MessageActionSheetMenuItemButton {
+  ShareButton({
+    super.key,
+    required super.message,
+    required super.bottomSheetContext,
+    required super.messageListContext,
+  });
+
+  @override IconData get icon => Icons.adaptive.share;
+
+  @override String get label => 'Share';
+
+  @override VoidCallback get onPressed => () async {
+    // Close the message action sheet; we're about to show the share
+    // sheet. (We could do this after the sharing Future settles, but
+    // on iOS I get impatient with how slowly our action sheet
+    // dismisses in that case.)
+    // TODO(#24): Fix iOS bug where this call causes the keyboard to
+    //   reopen (if it was open at the time of this
+    //   `showMessageActionSheet` call) and cover a large part of the
+    //   share sheet.
+    Navigator.of(bottomSheetContext).pop();
+
+    // TODO: to support iPads, we're asked to give a
+    //   `sharePositionOrigin` param, or risk crashing / hanging:
+    //     https://pub.dev/packages/share_plus#ipad
+    //   Perhaps a wart in the API; discussion:
+    //     https://github.com/zulip/zulip-flutter/pull/12#discussion_r1130146231
+    // TODO: Share raw Markdown, not HTML
+    await Share.shareWithResult(message.content);
+  };
+}
+
+class QuoteAndReplyButton extends MessageActionSheetMenuItemButton {
+  QuoteAndReplyButton({
+    super.key,
+    required super.message,
+    required super.bottomSheetContext,
+    required super.messageListContext,
+  });
+
+  @override IconData get icon => Icons.format_quote_outlined;
+
+  @override String get label => 'Quote and reply';
+
+  @override VoidCallback get onPressed => () async {
```

Comment:
There's a lot here in this method, and it'd be real nice to have a few tests for it. Would you try writing some?

I think after #30 we may have all the pieces we need in order to be able to write a reasonable test for code like this. If you run into something that makes it hard to test, let's discuss in chat.

Reply:
Sure! I'm happy to report that I've gotten a reasonable-looking "happy path" test in my next revision. It does these things:

- renders a message list with a message and a compose box
- simulates a long-press on the message to open the message action sheet
- simulates a tap on the "Quote and reply" button
- checks the compose-input value before and after the fetch-raw-Markdown request

It's 72 lines of code, which seems like a lot. I'd like to see how we could shape the code for good support of additional tests, such as to exercise when the fetch-raw-Markdown request fails. (And also I guess to check that we don't offer quote-and-reply at all when it can't be supported; i.e., when we don't offer a compose box for the narrow.) Maybe we can go over this in the office today?